modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +163,111 @@ impl VulnerabilityService {
 
         Ok(result.rows_affected)
     }
+
+    pub async fn analyze_purls<C>(
+        &self,
+        purls: Vec<&str>,
+        connection: &C,
+    ) -> Result<HashMap<String, Vec<VulnerabilityDetails>>, Error>
+    where
+        C: ConnectionTrait + StreamTrait,
+    {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let query = purls
+            .iter()
+            .filter_map(|p| {
+                let purl = Purl::from_str(p).map_err(Error::Purl);
+                let canonical = match purl {
+                    Ok(purl) => CanonicalPurl::from(purl),
+                    Err(_) => return None,
+                };
+
+                let version = match canonical.version.as_ref() {
+                    Some(v) => v,
+                    None => return None,
```

Comment:
Another silent fail? Personally, I'd change `filter_map` to `map` where the closure returns a `Result` that `collect` can turn into a `Vec` or return an error for any invalid purls or valid ones without a version.

Reply:
I'll update the response to return `vulnerabilities` and `errors`